design-proposals/preferences.md

```diff
@@ -0,0 +1,584 @@
+  Memory *v1.Memory `json:"memory,omitempty"`
+
+  // Machine type.
+  //
+  // +optional
+  Machine *v1.Machine `json:"machine,omitempty"`
+
+  // Firmware.
+  //
+  // +optional
+  Firmware *v1.Firmware `json:"firmware,omitempty"`
+
+  // Clock sets the clock and timers of the vmi.
+  //
+  // +optional
+  Clock *v1.Clock `json:"clock,omitempty"`
+
+  // Features like acpi, apic, hyperv, smm.
+  //
+  // +optional
+  Features *v1.Features `json:"features,omitempty"`
+
+  // Devices allows adding disks, network interfaces, and others
+  //
+  // +optional
+  Devices v1.Devices `json:"devices,omitempty"`
+
+  // Controls whether or not disks will share IOThreads.
+  // Omitting IOThreadsPolicy disables use of IOThreads.
+  // One of: shared, auto
+  //
+  // +optional
+  IOThreadsPolicy *v1.IOThreadsPolicy `json:"ioThreadsPolicy,omitempty"`
+
+  // Chassis specifies the chassis info passed to the domain.
+  //
+  // +optional
+  Chassis *v1.Chassis `json:"chassis,omitempty"`
+
+  // Launch Security setting of the vmi.
+  //
+  // +optional
+  LaunchSecurity *v1.LaunchSecurity `json:"launchSecurity,omitempty"`
+
+}
+```
+
+### `DevicesDefaults`, `DiskDefaults`, `InputDefaults` and `InterfaceDefaults`
+
+As the names suggest these types will provide preferences for `Devices` not present within `VirtualMachinePreferencesSpec` that could be present in the `VirtualMachine` definition and would then be applied to the `VirtualMachineInstance`.
+
+```go
+// DevicesDefaults contains various optional defaults for Disks, Interfaces and Inputs.
+//
+// +k8s:openapi-gen=true
+type DevicesDefaults struct {
+
+  // DiskDefaults optionally contains various preferred defaults for Disk Devices.
+  //
+  // +optional
+  DiskDefaults *DiskDefaults `json:"diskDefaults,omitempty"`
+
+  // InterfaceDefaults optionally contains various preferred defaults for Interface Devices.
+  //
+  // +optional
+  InterfaceDefaults *InterfaceDefaults `json:"interfaceDefaults,omitempty"`
+
+  // InputDefaults optionally contains various preferred defaults for Input Devices.
+  //
+  // +optional
+  InputDefaults *InputDefaults `json:"inputDefaults,omitempty"`
+
+}
+
+// DiskDefaults contains optional defaults for Disks
+//
+// +k8s:openapi-gen=true
+type DiskDefaults struct {
+
+  // PreferredDiskBus optionally defines the preferred bus for Disk Disk devices.
+  //
+  // +optional
+  PreferredDiskBus string `json:"preferredDiskBus,omitempty"`
+
+  // PreferredLunBus optionally defines the preferred bus for Lun Disk devices.
+  //
+  // +optional
+  PreferredLunBus string `json:"preferredLunBus,omitempty"`
+
+  // PreferredCdromBus optionally defines the preferred bus for Cdrom Disk devices.
+  //
+  // +optional
+  PreferredCdromBus string `json:"preferredCdromBus,omitempty"`
+
+  // PreferredDedicatedIoThread optionally enables dedicated IO threads for Disk devices.
+  //
+  // +optional
+  PreferredDedicatedIoThread *bool `json:"preferredDedicatedIoThread,omitempty"`
```

Comment:
Isn't this about the resources provided to the VM?

Reply:
Right I've left a TODO to determine which side things like this need to fall in the latest revision but I might expand this into a section covering it in a little more detail.